---
title: "Introduction to CASE for multi-cell-type eQTL fine-mapping"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Introduction_to_CASE}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# This Vignette is a draft and will be further updated.

First, we load necessary packages.

```{r setup}
library(CASE)
library(ggplot2)
library(tidyr)
set.seed(1000)
```

## The example data

The example genotype was generated by R package `CorBin`.

```{r}
data("example_data")
attach(example_data)

N = nrow(X)
M = ncol(X)
C = ncol(Y)
cat(" Sample size =", N, "\n",
    "SNP size =", M, "\n",
    "Cell type number =", C, "\n")
```

Let's have a look at the Linkage Disequilibrium (LD) of the example genotype.

```{r}
R = cor(X)

df <- expand.grid(seq(M), seq(M))
df$corr <- as.vector(R)

g2 = ggplot(df, aes(x = Var1, y = Var2, fill = corr)) +
  geom_tile() +
  scale_fill_gradient2(low = "blue", mid = "white", high = "red", limits = c(-1, 1),
                       midpoint = 0, na.value = NA,
                       guide = guide_colorbar(
                         barwidth = 1, barheight = 7, title.position = "top", oob = scales::oob_squish
                       )
                       ) +
  xlab("") + 
  ylab("") +
  theme(aspect.ratio = 1, axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
        strip.background = element_rect(fill = "white", color = "white"),
        panel.border = element_rect(colour = "black", fill = NA, 
                                    linewidth = 1),
        strip.text = element_text(color = "black", size = 20),
        text = element_text(size = 20)) +
  labs(fill = "corr")
print(g2)
```

Only SNP 10 and SNP 950 have eQTL effects for all three cell types and for the first two cell types, respectively. 

```{r}
print(which(B != 0, arr.ind = TRUE))
idx1 = 10
idx2 = 950
B[c(idx1, idx2), ]
```
Check the Z scores for the causal SNPs.

```{r}
Z = matrix(0, M, C)
for (i in 1:M){
  m1 = summary(lm(Y ~ X[, i]))
  Z[i, ] = sapply(m1, function(x) x$coefficients[2, 3])
}
Z[c(idx1, idx2), ]
```

Visualization of the Z scores for three cell types. The true eQTLs were surrounded with green circles. The dashed grey and red lines are for nominal significance ($\pm 1.96$) and significance after Bonferroni adjustment ($\pm 4.055$).

```{r}
df <- as.data.frame(Z)
colnames(df) = paste0("Cell_type_", 1:3)
df$SNP <- 1:nrow(df)

df_long <- pivot_longer(df, 
                        cols = starts_with("Cell_type"), 
                        names_to = "Variable", 
                        values_to = "Z")
df_long$Highlight <- with(df_long, 
  (SNP == 10) | 
  (SNP == 950 & Variable != "Cell_type_3")
)

ggplot(df_long, aes(x = SNP, y = Z)) +
  geom_point() +
  geom_point(data = subset(df_long, Highlight), 
             aes(x = SNP, y = Z), 
             color = "green", size = 4, shape = 1) +
  geom_hline(yintercept = c(1.96, -1.96), linetype = "dashed", color = "grey") +
  geom_hline(yintercept = c(4.055, -4.055), linetype = "dashed", color = "red") +
  facet_wrap(~ Variable) + 
  theme_minimal()
```

## Fine-mapping

First, we try CASE fine-mapping for each cell type separately. The results lack power for the third cell type and for SNP 950.

```{r}
for (c in 1:C){
  f1 = CASE(Z = Z[, c], R = R, N = N)
  print(f1$sets)
}
```

However, CASE jointly studies three cell types together to improve the power of identifying the causal variants.

```{r}
fit <- CASE(Z = Z, R = R, N = N)
print(fit$sets)
```

## Session Information

```{r}
sessionInfo()
```
